worlds/kh2/Rules.py

```diff
@@ -1,96 +1,1839 @@
+        """
+        return any(
+                [self.kh2_can_reach(location, state) for location in
+                 loc_set]
+        )
+
+    def kh2_can_reach_all(self, loc_list: list, state: CollectionState):
+        """
+        Can reach all locations in the set.
+        """
+        return all(
+                [state.can_reach(self.world.multiworld.get_location(location, self.player), "location", self.player) for location in
+                 loc_list]
+        )
+
+    def kh2_can_reach(self, loc: str, state: CollectionState):
+        """
+        Returns bool instead of collection state.
+        """
+        if state.can_reach(self.world.multiworld.get_location(loc, self.player), "location", self.player):
+            return True
         else:
-            world.completion_condition[player] = lambda state: state.kh_hitlist(player, world.BountyRequired[player].value)
+            return False
+
+    def kh2_has_all(self, items: list, state: CollectionState):
+        """If state has at least one of all."""
+        return state.has_all(set(items), self.player)
+
+    def kh2_has_any(self, items: list, state: CollectionState):
+        return state.has_any(set(items), self.player)
+
+    def drive_form_unlock(self, state: CollectionState, parent_drive_form, level_required, fight_logic=False) -> bool:
+        form_access = {parent_drive_form}
+        if self.world.multiworld.AutoFormLogic[self.player] and state.has(ItemName.SecondChance, self.player) and not fight_logic:
+            if parent_drive_form == ItemName.MasterForm:
+                if state.has(ItemName.DriveConverter, self.player):
+                    form_access.add(self.auto_form_dict[parent_drive_form])
+            else:
+                form_access.add(self.auto_form_dict[parent_drive_form])
+        return state.has_any(form_access, self.player) \
+            and self.get_form_level_requirement(state, level_required)
+
+    def get_form_level_requirement(self, state, amount):
+        forms_available = 0
+        form_list = [ItemName.ValorForm, ItemName.WisdomForm, ItemName.LimitForm, ItemName.MasterForm,
+                     ItemName.FinalForm]
+        if self.world.multiworld.FinalFormLogic[self.player] != "no_light_and_darkness":
+            if self.world.multiworld.FinalFormLogic[self.player] == "light_and_darkness":
+                if state.has(ItemName.LightDarkness, self.player) and state.has_any(set(form_list), self.player):
+                    forms_available += 1
+                    form_list.remove(ItemName.FinalForm)
+
+            else:  # self.multiworld.FinalFormLogic=="just a form"
+                form_list.remove(ItemName.FinalForm)
+                if state.has_any(form_list, self.player):
+                    forms_available += 1
+        forms_available += sum([1 for form in form_list if state.has(form, self.player)])
+        return forms_available >= amount
+
+
+class KH2FormRules(KH2Rules):
```

Comment:
the reason i inherit the way i do in messenger is because only a single one of those classes ever actually gets instantiated. you're instantiating every single one here, so every single thing defined in KH2Rules is also being defined here which is incredibly expensive. you should make a barebones KH2Rules that just consists of the helpers, and then inherit everything else from that.